### ISSUE TITLE
nix-build works locally, but not in docker

### ISSUE THREAD

**Issue by Staff-d** [NixOS/docker#26]
Good day everyone,

I use the `nixos/nix` image in my ci pipeline and am currently trying to build a derivation which was created by [node2nix](https://github.com/svanderburg/node2nix). The derivation works fine on my machine and any other I could try, but stops working when used inside the `nixos/nix` docker image. The derivation "succeeds" and creates an output both locally and inside the docker, but all of my unittests are failing in the docker container as all binary dependencies are missing from the environment. 

After a long session of trial and error, I created a graph of the dependencies of the derivation both locally and in the docker, using `nix-store -q --graph ... | dot`. 

A comparison of these graphs can be found below. The graph generated inside docker is missing most of the dependencies:

Graph created inside the docker image:  
![dockerOut](https://user-images.githubusercontent.com/8281040/94932678-b2d25b00-04c9-11eb-80e4-2eaedbaf1c51.png)

Graph created locally:
![localOut](https://user-images.githubusercontent.com/8281040/94932690-b6fe7880-04c9-11eb-9e86-f949c5645282.png)

Things I have tried to locate the source of the problem:

- Spawned a fresh Ubuntu 20 VM, installed nix and executed `nix-build`. The derivation works.
- Spawned a fresh NixOS 20.03 VM and executed `nix-build`. The derivation works.
- Tried the derivation on two other physical machines. It derivation works.
- I executed `nix-build -vvvvv` on my machine and in the docker, and `git diff`ed the ouput. They are equal (except for the paths of the mountpoint and runner ids).
- I added a `printenv` in the build step and, again, `git diff`ed the output. Same results as above.
- I have updated the version of nix inside the container to `2.3.7` (the same version I use locally). Same results.

I have created a small [repository](https://github.com/Staff-d/node2nix-docker-bug) containing the necessary code to reproduce my problem. This repository is missing my failing unit-tests, but the store graph is reproducible this way.

I have to admit that I am totally at loss here, as this behaviour goes against my understanding of how nix works. If I have made any noobie mistakes, please feel free to point them out.

Best regards:
Sebastian

**Comment by Staff-d** [comment]
So, after I set this problem aside for a few days I came back to it yesterday and looked at the output again.

I already realized before that the missing dependencies later down the chain are caused by the shrinking of the RPATH variable. On my local machine it only strips a few debug libraries, but inside the container the output looks like this

```
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libgif.so.7
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libgthread-2.0.so.0
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libpangocairo-1.0.so.0
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libstdc++.so.6
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libcairo.so.2
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libgmodule-2.0.so.0
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libgobject-2.0.so.0
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libfreetype.so.6
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libpcre.so.1
shrinking /nix/store/*truncated*/lib/node_modules/canvas/build/Release/libfribidi.so.0
```
Most of the dependencies (`pango`,`cairo`, ...) I added are being removed again. But why?

I ran `nix-build` again with slightly less verbosity this time and were able to spot where the differences in the build started this time. `npm install` tries to fetch the dependencies for the `canvas` module. On my machine the relevant lines look like this:

```
canvas@2.6.1 install /nix/store/8kla2yf4mzcdw2d5nzg8pfydhnsa2smz-node-dependencies-Node2NixProblem-0.0.1/Node2NixProblem/node_modules/canvas
node-pre-gyp install --fallback-to-build
node-pre-gyp WARN Using needle for node-pre-gyp https download 
node-pre-gyp WARN Pre-built binaries not installable for canvas@2.6.1 and node@12.18.2 (node-v72 ABI, glibc) (falling back to source compile with node-gyp) 
node-pre-gyp WARN Hit error getaddrinfo ENOTFOUND github.com 
//...
//node continues to build canvas from source
```
but inside the docker node is able to fetch the prebuild binary from github
```
canvas@2.6.1 install /nix/store/nwp3rplg6w55lj6symxrl2w07cn473ww-node-dependencies-Node2NixProblem-0.0.1/Node2NixProblem/node_modules/canvas
node-pre-gyp install --fallback-to-build
node-pre-gyp WARN Using needle for node-pre-gyp https download 
download thread waiting for 10000 ms
[canvas] Success: "/nix/store/nwp3rplg6w55lj6symxrl2w07cn473ww-node-dependencies-Node2NixProblem-0.0.1/Node2NixProblem/node_modules/canvas/build/Release/canvas.node" is installed via remote
//...
//node continues with other dependencies
```
When the dependency is sideloaded, my inputs do not get used and will be (rightfully) stripped from RPATH and are missing later (when they are needed on runtime). 

Why sideloading is allowed/works inside the docker in the first place is something I don't understand. Sandboxing is (according to the [current manual](https://github.com/NixOS/nix/blob/9b9d5297baf02923a1e2879441bc1e25190d01cc/doc/manual/command-ref/conf-file.xml#L807)) turned on by default. The dockerfile turns [sandboxing off](https://github.com/NixOS/docker/blob/master/Dockerfile#L16), but even when I delete that line, build a new docker image and run my tests again the dependency still gets sideloaded.

Does anyone know what causes this behaviour?

**Comment by Staff-d** [comment]
Okay, I found the error.
For sandboxing to work, the docker must be started with [`--privileged`](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities), as ["documented"](https://github.com/NixOS/docker/commit/fd6472552945aef2fced5f0d1bddbfe752188423) in this commit message.

I will close this issue know and create a PR to update the readme with this information to save other people the hassle I went through.
